File: Day 25/PathSum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        node_sum = 0
        if root == None:
            return False
        return self.findSum(root,sum,node_sum)
        
    def findSum(self,root,given_sum,n_sum):
        n_sum += root.val
        final_res = False
        
        if root.left == None and root.right == None:
            if n_sum == given_sum:
                print(n_sum)
                return True
            return False
        
        if root.left != None:
            l_res = self.findSum(root.left,given_sum,n_sum)
            if l_res:
                final_res = l_res
        if root.right != None:
            r_res = self.findSum(root.right,given_sum,n_sum)
            if r_res:
                final_res = r_res
        
        return final_res

File: Round 2 Preparation/Popular Google Questions/PathSum.java
//https://leetcode.com/problems/path-sum-iii/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int sum) {
        if(root == null){
            return 0;
        }
        LinkedList<TreeNode> queue = new LinkedList<>();
        int pathSum = 0;
        queue.add(root);
        int counter = 0;
        while(queue.size() != 0){
            TreeNode rootPopped = queue.remove();
            int tempSum =  getSum(rootPopped,0,sum);
            pathSum += tempSum;
            //System.out.println(rootPopped.val + " > " +tempSum);
            if(rootPopped.left != null){
                //System.out.println(rootPopped.left.val);
                queue.add(rootPopped.left);
            }
            if(rootPopped.right != null){
                //System.out.println(rootPopped.right.val);
                queue.add(rootPopped.right);
            }
           
        }
        return pathSum;       
    }
  
    public int getSum(TreeNode root,int sum,int target){
            if(root.left == null && root.right == null){
                sum += root.val;
                if(sum == target){
                    return 1;
                }
                return 0;
            }
            sum += root.val;
            int resultSum = 0;
            if(sum == target){
                resultSum++;
            }
            
            if(root.left != null){
               resultSum += getSum(root.left, sum, target); 
            }

            if(root.right != null){
                resultSum += getSum(root.right, sum, target);
            }
        return resultSum;
    }
        
        
}